### PR TITLE
Fix nll_backward for negative weights.

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64275

Fixes https://github.com/pytorch/pytorch/issues/64256
It adds a test with negative weights as well.

This PR also cleans the code of `nll_backward` which contained a number
of inconsistencies between the different branches of the code: some of
them would check that an index was correct while others didn't. Now the
code is more homogeneous and hopefully easier to follow.

### CODE REVIEW

**Comment by Lezcano** [comment]
cc @pmeier

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -5271,6 +5271,7 @@ def sample_inputs_nll_loss(op_info, device, dtype, requires_grad, **kwargs):
         ((*shape, 1, 2), dict()),
         ((*shape, 1, 2, 3), dict()),
         (shape, dict(weight=make_tensor((num_classes,), device=device, dtype=dtype).abs())),
+        (shape, dict(weight=make_tensor((num_classes,), device=device, dtype=dtype))),
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:5274]: If we explicitly want to test this for positive and negative weights, maybe we should simply set `low=0` (and remove the `.abs()`) in the line above and here set 

```suggestion
        (shape, dict(weight=make_tensor((num_classes,), high=0, device=device, dtype=dtype))),
```

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py]: Good point. Corrected!

Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py]: Instead of leaving the `.abs()` call in the line above, could you just set `low=0` in the `make_tensor()` call? Or remove it completely to indicate "any kind of weight".

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:5275]: Left one with positive, one with negative and one with mixed.

**Review by jbschlosser — CHANGES_REQUESTED** [review]
Thanks @Lezcano for the fix! It looks like Jacobian checks are still failing on CUDA - guess we need similar changes in `native/cuda/Loss.cu` as well

**Comment by Lezcano** [comment]
You are completely right @jbschlosser ! It should be fixed now.

I also removed some guards that did not give the right result when the sum of the weights was zero. If the sum of the passed weights is zero and the reduction is `mean`, this operation should return infinities, and same for the gradients. Otherwise it's silently returning a finite but incorrect result.

**Comment by Lezcano** [comment]
Oh, sorry, the CI's failing. I'll fix it on Monday first thing.

**Comment by Lezcano** [comment]
I believe this should be ready for review @jbschlosser

**Comment by Lezcano** [comment]
Actually, I've realised that this is a linear function, and as such, the derivative should be implemented in terms of the forward (modulo saving some precomputations from the forward). I'm going to scrap what I have and factorize the code so that the forward and the backward use basically the same implementation.
